File: src/services/weatherPatternAnalyzer.js
/**
 * Weather Pattern Analyzer
 * Analyzes weather patterns, temperature ranges, and precipitation data
 * to identify trends and make intelligent predictions
 */

export class WeatherPatternAnalyzer {
  constructor() {
    this.historicalData = this.loadHistoricalData();
    this.patterns = {
      temperatureTrends: [],
      precipitationPatterns: [],
      seasonalData: {},
    };
  }

  /**
   * Analyze weather data to extract meaningful patterns
   */
  analyzeWeatherData(currentWeather, forecast) {
    const analysis = {
      currentConditions: this.analyzeCurrentConditions(currentWeather),
      temperatureTrend: this.analyzeTemperatureTrend(forecast),
      precipitationRisk: this.analyzePrecipitationRisk(currentWeather, forecast),
      comfortIndex: this.calculateComfortIndex(currentWeather),
      weatherStability: this.calculateWeatherStability(forecast),
      timeOfDay: this.getTimeOfDayContext(),
    };

    // Store for historical learning
    this.storeWeatherPattern(analysis);

    return analysis;
  }

  /**
   * Analyze current weather conditions with detailed metrics
   */
  analyzeCurrentConditions(weather) {
    const temp = weather.current.temperature;
    const condition = (weather.current.condition || '').toLowerCase();

    return {
      temperature: temp,
      feelsLike: this.calculateFeelsLike(temp, weather.current),
      tempCategory: this.categorizeTemperature(temp),
      condition: condition,
      isRainy: condition.includes('rain') || condition.includes('drizzle'),
      isSnowy: condition.includes('snow'),
      isCloudy: condition.includes('cloud'),
      isSunny: condition.includes('clear') || condition.includes('sun'),
      humidity: weather.current.humidity || 50,
      windSpeed: weather.current.wind?.speed || 0,
    };
  }

  /**
   * Analyze temperature trends from forecast data
   */
  analyzeTemperatureTrend(forecast) {
    if (!forecast || forecast.length < 2) {
      return { trend: 'stable', change: 0, volatility: 'low' };
    }

    const temps = forecast.map(f => f.temp);
    const changes = [];

    for (let i = 1; i < temps.length; i++) {
      changes.push(temps[i] - temps[i - 1]);
    }

    const avgChange = changes.reduce((a, b) => a + b, 0) / changes.length;
    const volatility = this.calculateVolatility(changes);

    return {
      trend: avgChange > 2 ? 'warming' : avgChange < -2 ? 'cooling' : 'stable',
      change: avgChange,
      volatility: volatility > 5 ? 'high' : volatility > 2 ? 'medium' : 'low',
      maxTemp: Math.max(...temps),
      minTemp: Math.min(...temps),
      range: Math.max(...temps) - Math.min(...temps),
    };
  }

  /**
   * Analyze precipitation risk based on current and forecast data
   */
  analyzePrecipitationRisk(currentWeather, forecast) {
    const currentCondition = (currentWeather.current.condition || '').toLowerCase();
    const isCurrentlyRainy = currentCondition.includes('rain') ||
                            currentCondition.includes('drizzle') ||
                            currentCondition.includes('snow');

    const forecastRainCount = forecast.filter(f => {
      const cond = (f.condition || '').toLowerCase();
      return cond.includes('rain') || cond.includes('drizzle') || cond.includes('snow');
    }).length;

    const riskScore = (isCurrentlyRainy ? 0.5 : 0) +
                     (forecastRainCount / forecast.length * 0.5);

    return {
      level: riskScore > 0.6 ? 'high' : riskScore > 0.3 ? 'medium' : 'low',
      score: riskScore,
      isCurrentlyRainy,
      forecastRainProbability: forecastRainCount / forecast.length,
      type: currentCondition.includes('snow') ? 'snow' :
            currentCondition.includes('rain') ? 'rain' : 'none',
    };
  }

  /**
   * Calculate comfort index based on temperature, humidity, and wind
   */
  calculateComfortIndex(weather) {
    const temp = weather.current.temperature;
    const humidity = weather.current.humidity || 50;
    const windSpeed = weather.current.wind?.speed || 0;

    // Heat index calculation (simplified)
    let comfortScore = 50;

    if (temp < 10) comfortScore -= (10 - temp) * 2;
    else if (temp > 28) comfortScore -= (temp - 28) * 2;

    if (humidity > 70) comfortScore -= (humidity - 70) * 0.5;
    if (windSpeed > 20) comfortScore -= (windSpeed - 20) * 0.3;

    return {
      score: Math.max(0, Math.min(100, comfortScore)),
      level: comfortScore > 70 ? 'comfortable' :
             comfortScore > 40 ? 'moderate' : 'uncomfortable',
      factors: {
        temperature: temp > 18 && temp < 26 ? 'ideal' : 'suboptimal',
        humidity: humidity > 30 && humidity < 60 ? 'ideal' : 'suboptimal',
        wind: windSpeed < 15 ? 'calm' : 'breezy',
      },
    };
  }

  /**
   * Calculate weather stability from forecast
   */
  calculateWeatherStability(forecast) {
    if (!forecast || forecast.length < 2) return { stable: true, score: 1 };

    const conditions = forecast.map(f => f.condition);
    const uniqueConditions = new Set(conditions).size;
    const stabilityScore = 1 - (uniqueConditions / conditions.length);

    return {
      stable: stabilityScore > 0.6,
      score: stabilityScore,
      level: stabilityScore > 0.7 ? 'very stable' :
             stabilityScore > 0.4 ? 'moderately stable' : 'unstable',
    };
  }

  /**
   * Get time of day context for recommendations
   */
  getTimeOfDayContext() {
    const hour = new Date().getHours();

    return {
      hour,
      period: hour < 6 ? 'night' :
              hour < 12 ? 'morning' :
              hour < 17 ? 'afternoon' :
              hour < 21 ? 'evening' : 'night',
      isWorkHours: hour >= 9 && hour < 17,
      needsAllDayGear: hour < 10, // Early enough to plan for whole day
    };
  }

  /**
   * Calculate "feels like" temperature
   */
  calculateFeelsLike(temp, current) {
    const windSpeed = current.wind?.speed || 0;
    const humidity = current.humidity || 50;

    // Wind chill for cold weather
    if (temp < 10 && windSpeed > 5) {
      return temp - (windSpeed * 0.5);
    }

    // Heat index for hot weather
    if (temp > 27 && humidity > 40) {
      return temp + ((humidity - 40) * 0.1);
    }

    return temp;
  }

  /**
   * Categorize temperature into ranges
   */
  categorizeTemperature(temp) {
    if (temp < 0) return 'freezing';
    if (temp < 5) return 'very_cold';
    if (temp < 10) return 'cold';
    if (temp < 15) return 'cool';
    if (temp < 20) return 'mild';
    if (temp < 25) return 'warm';
    if (temp < 30) return 'hot';
    return 'very_hot';
  }

  /**
   * Calculate volatility (standard deviation)
   */
  calculateVolatility(values) {
    const mean = values.reduce((a, b) => a + b, 0) / values.length;
    const variance = values.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / values.length;
    return Math.sqrt(variance);
  }

  /**
   * Store weather pattern for historical analysis
   */
  storeWeatherPattern(analysis) {
    this.historicalData.push({
      timestamp: Date.now(),
      ...analysis,
    });

    // Keep only last 100 entries
    if (this.historicalData.length > 100) {
      this.historicalData = this.historicalData.slice(-100);
    }

    this.saveHistoricalData();
  }

  /**
   * Load historical data from localStorage
   */
  loadHistoricalData() {
    try {
      const data = localStorage.getItem('weather_historical_data');
      return data ? JSON.parse(data) : [];
    } catch (error) {
      console.error('Error loading historical data:', error);
      return [];
    }
  }

  /**
   * Save historical data to localStorage
   */
  saveHistoricalData() {
    try {
      localStorage.setItem('weather_historical_data', JSON.stringify(this.historicalData));
    } catch (error) {
      console.error('Error saving historical data:', error);
    }
  }

  /**
   * Get seasonal patterns from historical data
   */
  getSeasonalPatterns() {
    const month = new Date().getMonth();
    const season = Math.floor(month / 3); // 0=winter, 1=spring, 2=summer, 3=fall

    const seasonalData = this.historicalData.filter(entry => {
      const entryMonth = new Date(entry.timestamp).getMonth();
      return Math.floor(entryMonth / 3) === season;
    });

    if (seasonalData.length < 5) {
      return null; // Not enough data
    }

    const avgTemp = seasonalData.reduce((sum, entry) =>
      sum + entry.currentConditions.temperature, 0) / seasonalData.length;

    const rainFrequency = seasonalData.filter(entry =>
      entry.precipitationRisk.isCurrentlyRainy).length / seasonalData.length;

    return {
      season: ['winter', 'spring', 'summer', 'fall'][season],
      avgTemperature: avgTemp,
      rainFrequency,
      dataPoints: seasonalData.length,
    };
  }
}

File: backend/app/services/weather_pattern_analyzer.py
"""
Weather Pattern Analyzer
Analyzes weather patterns, temperature ranges, and precipitation data
to identify trends and make intelligent predictions
"""
import math
from datetime import datetime
from typing import Dict, List, Any, Optional


class WeatherPatternAnalyzer:
    def __init__(self, historical_data: List[Dict] = None):
        self.historical_data = historical_data or []
        self.patterns = {
            'temperatureTrends': [],
            'precipitationPatterns': [],
            'seasonalData': {}
        }

    def analyze_weather_data(
        self,
        current_weather: Dict[str, Any],
        forecast: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Analyze weather data to extract meaningful patterns"""
        analysis = {
            'currentConditions': self.analyze_current_conditions(current_weather),
            'temperatureTrend': self.analyze_temperature_trend(forecast),
            'precipitationRisk': self.analyze_precipitation_risk(current_weather, forecast),
            'comfortIndex': self.calculate_comfort_index(current_weather),
            'weatherStability': self.calculate_weather_stability(forecast),
            'timeOfDay': self.get_time_of_day_context()
        }

        # Store for historical learning
        self.store_weather_pattern(analysis)

        return analysis

    def analyze_current_conditions(self, weather: Dict[str, Any]) -> Dict[str, Any]:
        """Analyze current weather conditions with detailed metrics"""
        current = weather['current']
        temp = current['temperature']
        condition = (current.get('condition', '')).lower()

        return {
            'temperature': temp,
            'feelsLike': self.calculate_feels_like(temp, current),
            'tempCategory': self.categorize_temperature(temp),
            'condition': condition,
            'isRainy': 'rain' in condition or 'drizzle' in condition,
            'isSnowy': 'snow' in condition,
            'isCloudy': 'cloud' in condition,
            'isSunny': 'clear' in condition or 'sun' in condition,
            'humidity': current.get('humidity', 50),
            'windSpeed': current.get('wind', {}).get('speed', 0)
        }

    def analyze_temperature_trend(
        self,
        forecast: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Analyze temperature trends from forecast data"""
        if not forecast or len(forecast) < 2:
            return {'trend': 'stable', 'change': 0, 'volatility': 'low'}

        temps = [f['temp'] for f in forecast]
        changes = [temps[i] - temps[i - 1] for i in range(1, len(temps))]

        avg_change = sum(changes) / len(changes)
        volatility = self.calculate_volatility(changes)

        return {
            'trend': 'warming' if avg_change > 2 else 'cooling' if avg_change < -2 else 'stable',
            'change': avg_change,
            'volatility': 'high' if volatility > 5 else 'medium' if volatility > 2 else 'low',
            'maxTemp': max(temps),
            'minTemp': min(temps),
            'range': max(temps) - min(temps)
        }

    def analyze_precipitation_risk(
        self,
        current_weather: Dict[str, Any],
        forecast: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Analyze precipitation risk based on current and forecast data"""
        current_condition = (current_weather['current'].get('condition', '')).lower()
        is_currently_rainy = ('rain' in current_condition or
                              'drizzle' in current_condition or
                              'snow' in current_condition)

        forecast_rain_count = sum(
            1 for f in forecast
            if 'rain' in (f.get('condition', '')).lower() or
               'drizzle' in (f.get('condition', '')).lower() or
               'snow' in (f.get('condition', '')).lower()
        )

        risk_score = (0.5 if is_currently_rainy else 0) + (forecast_rain_count / len(forecast) * 0.5)

        return {
            'level': 'high' if risk_score > 0.6 else 'medium' if risk_score > 0.3 else 'low',
            'score': risk_score,
            'isCurrentlyRainy': is_currently_rainy,
            'forecastRainProbability': forecast_rain_count / len(forecast) if forecast else 0,
            'type': 'snow' if 'snow' in current_condition else 'rain' if 'rain' in current_condition else 'none'
        }

    def calculate_comfort_index(self, weather: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate comfort index based on temperature, humidity, and wind"""
        current = weather['current']
        temp = current['temperature']
        humidity = current.get('humidity', 50)
        wind_speed = current.get('wind', {}).get('speed', 0)

        # Heat index calculation (simplified)
        comfort_score = 50

        if temp < 10:
            comfort_score -= (10 - temp) * 2
        elif temp > 28:
            comfort_score -= (temp - 28) * 2

        if humidity > 70:
            comfort_score -= (humidity - 70) * 0.5
        if wind_speed > 20:
            comfort_score -= (wind_speed - 20) * 0.3

        comfort_score = max(0, min(100, comfort_score))

        return {
            'score': comfort_score,
            'level': 'comfortable' if comfort_score > 70 else 'moderate' if comfort_score > 40 else 'uncomfortable',
            'factors': {
                'temperature': 'ideal' if 18 < temp < 26 else 'suboptimal',
                'humidity': 'ideal' if 30 < humidity < 60 else 'suboptimal',
                'wind': 'calm' if wind_speed < 15 else 'breezy'
            }
        }

    def calculate_weather_stability(
        self,
        forecast: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Calculate weather stability from forecast"""
        if not forecast or len(forecast) < 2:
            return {'stable': True, 'score': 1}

        conditions = [f.get('condition', '') for f in forecast]
        unique_conditions = len(set(conditions))
        stability_score = 1 - (unique_conditions / len(conditions))

        return {
            'stable': stability_score > 0.6,
            'score': stability_score,
            'level': 'very stable' if stability_score > 0.7 else 'moderately stable' if stability_score > 0.4 else 'unstable'
        }

    def get_time_of_day_context(self) -> Dict[str, Any]:
        """Get time of day context for recommendations"""
        hour = datetime.now().hour

        if hour < 6:
            period = 'night'
        elif hour < 12:
            period = 'morning'
        elif hour < 17:
            period = 'afternoon'
        elif hour < 21:
            period = 'evening'
        else:
            period = 'night'

        return {
            'hour': hour,
            'period': period,
            'isWorkHours': 9 <= hour < 17,
            'needsAllDayGear': hour < 10
        }

    def calculate_feels_like(self, temp: float, current: Dict[str, Any]) -> float:
        """Calculate 'feels like' temperature"""
        wind_speed = current.get('wind', {}).get('speed', 0)
        humidity = current.get('humidity', 50)

        # Wind chill for cold weather
        if temp < 10 and wind_speed > 5:
            return temp - (wind_speed * 0.5)

        # Heat index for hot weather
        if temp > 27 and humidity > 40:
            return temp + ((humidity - 40) * 0.1)

        return temp

    @staticmethod
    def categorize_temperature(temp: float) -> str:
        """Categorize temperature into ranges"""
        if temp < 0:
            return 'freezing'
        elif temp < 5:
            return 'very_cold'
        elif temp < 10:
            return 'cold'
        elif temp < 15:
            return 'cool'
        elif temp < 20:
            return 'mild'
        elif temp < 25:
            return 'warm'
        elif temp < 30:
            return 'hot'
        else:
            return 'very_hot'

    @staticmethod
    def calculate_volatility(values: List[float]) -> float:
        """Calculate volatility (standard deviation)"""
        if not values:
            return 0

        mean = sum(values) / len(values)
        variance = sum((val - mean) ** 2 for val in values) / len(values)
        return math.sqrt(variance)

    def store_weather_pattern(self, analysis: Dict[str, Any]) -> None:
        """Store weather pattern for historical analysis"""
        self.historical_data.append({
            'timestamp': datetime.now().timestamp() * 1000,
            **analysis
        })

        # Keep only last 100 entries
        if len(self.historical_data) > 100:
            self.historical_data = self.historical_data[-100:]

    def get_seasonal_patterns(self) -> Optional[Dict[str, Any]]:
        """Get seasonal patterns from historical data"""
        month = datetime.now().month
        season = month // 3  # 0=winter, 1=spring, 2=summer, 3=fall

        seasonal_data = [
            entry for entry in self.historical_data
            if datetime.fromtimestamp(entry['timestamp'] / 1000).month // 3 == season
        ]

        if len(seasonal_data) < 5:
            return None  # Not enough data

        avg_temp = sum(
            entry['currentConditions']['temperature']
            for entry in seasonal_data
        ) / len(seasonal_data)

        rain_frequency = sum(
            1 for entry in seasonal_data
            if entry['precipitationRisk']['isCurrentlyRainy']
        ) / len(seasonal_data)

        return {
            'season': ['winter', 'spring', 'summer', 'fall'][season],
            'avgTemperature': avg_temp,
            'rainFrequency': rain_frequency,
            'dataPoints': len(seasonal_data)
        }
